In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv('Shopping_Mart.csv')

In [3]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

     InvoiceDate  UnitPrice  CustomerID  
0  01/12/10 8:26       2.55     17850.0  
1  01/12/10 8:26       3.39     17850.0  
2  01/12/10 8:26       2.75     17850.0  
3  01/12/10 8:26       3.39     17850.0  
4  01/12/10 8:26       3.39     17850.0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495478 entries, 0 to 495477
Data columns (total 7 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    495478 non-null  object 
 1   StockCode    495478 non-null  object 
 2   Description  494024 non-null  object 
 3   Quantity     495478 non-null  int64  
 4   InvoiceDate  495478 non-null  object 
 5   UnitPrice    495478 non-null  float64
 6   CustomerID   361878 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 26.5+ MB


In [5]:
df.shape

(495478, 7)

In [6]:
data = df.groupby(['InvoiceNo', 'Description'])['Quantity'].sum()
data

InvoiceNo  Description                        
536365     CREAM CUPID HEARTS COAT HANGER          8
           GLASS STAR FROSTED T-LIGHT HOLDER       6
           KNITTED UNION FLAG HOT WATER BOTTLE     6
           RED WOOLLY HOTTIE WHITE HEART.          6
           SET 7 BABUSHKA NESTING BOXES            2
                                                  ..
C581490    ZINC T-LIGHT HOLDER STARS SMALL       -11
C581499    Manual                                 -1
C581568    VICTORIAN SEWING BOX LARGE             -5
C581569    36 PENCILS TUBE RED RETROSPOT          -5
           HANGING HEART JAR T-LIGHT HOLDER       -1
Name: Quantity, Length: 483378, dtype: int64

In [7]:
data = data.unstack().reset_index()

In [8]:
data

Description InvoiceNo   4 PURPLE FLOCK DINNER CANDLES  \
0              536365                             NaN   
1              536366                             NaN   
2              536367                             NaN   
3              536368                             NaN   
4              536369                             NaN   
...               ...                             ...   
22035         C581484                             NaN   
22036         C581490                             NaN   
22037         C581499                             NaN   
22038         C581568                             NaN   
22039         C581569                             NaN   

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
0                                       NaN                 NaN   
1                                       NaN                 NaN   
2                                       NaN                 NaN   
3                                       NaN                 NaN   
4                                       NaN                 NaN   
...                                     ...                 ...   
22035                                   NaN                 NaN   
22036                                   NaN                 NaN   
22037                                   NaN                 NaN   
22038                                   NaN                 NaN   
22039                                   NaN                 NaN   

Description   I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
0                                     NaN                       NaN   
1                                     NaN                       NaN   
2                                     NaN                       NaN   
3                                     NaN                       NaN   
4                                     NaN                       NaN   
...                                   ...                       ...   
22035                                 NaN                       NaN   
22036                                 NaN                       NaN   
22037                                 NaN                       NaN   
22038                                 NaN                       NaN   
22039                                 NaN                       NaN   

Description   OVAL WALL MIRROR DIAMANTE    RED SPOT GIFT BAG LARGE  \
0                                    NaN                       NaN   
1                                    NaN                       NaN   
2                                    NaN                       NaN   
3                                    NaN                       NaN   
4                                    NaN                       NaN   
...                                  ...                       ...   
22035                                NaN                       NaN   
22036                                NaN                       NaN   
22037                                NaN                       NaN   
22038                                NaN                       NaN   
22039                                NaN                       NaN   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  ...  \
0                                         NaN                      NaN  ...   
1                                         NaN                      NaN  ...   
2                                         NaN                      NaN  ...   
3                                         NaN                      NaN  ...   
4                                         NaN                      NaN  ...   
...                                       ...                      ...  ...   
22035                                     NaN                      NaN  ...   
22036                                     NaN                      NaN  ...   
22037                                     NaN                      NaN  ...   
22038                     

In [9]:
data = data.fillna(0).set_index('InvoiceNo')
data.head()

Description   4 PURPLE FLOCK DINNER CANDLES   50'S CHRISTMAS GIFT BAG LARGE  \
InvoiceNo                                                                     
536365                                  0.0                             0.0   
536366                                  0.0                             0.0   
536367                                  0.0                             0.0   
536368                                  0.0                             0.0   
536369                                  0.0                             0.0   

Description   DOLLY GIRL BEAKER   I LOVE LONDON MINI BACKPACK  \
InvoiceNo                                                       
536365                      0.0                           0.0   
536366                      0.0                           0.0   
536367                      0.0                           0.0   
536368                      0.0                           0.0   
536369                      0.0                           0.0   

Description   NINE DRAWER OFFICE TIDY   OVAL WALL MIRROR DIAMANTE   \
InvoiceNo                                                            
536365                            0.0                          0.0   
536366                            0.0                          0.0   
536367                            0.0                          0.0   
536368                            0.0                          0.0   
536369                            0.0                          0.0   

Description   RED SPOT GIFT BAG LARGE   SET 2 TEA TOWELS I LOVE LONDON   \
InvoiceNo                                                                 
536365                            0.0                               0.0   
536366                            0.0                               0.0   
536367                            0.0                               0.0   
536368                            0.0                               0.0   
536369                            0.0                               0.0   

Description   SPACEBOY BABY GIFT SET   TOADSTOOL BEDSIDE LIGHT   ...  \
InvoiceNo                                                        ...   
536365                           0.0                        0.0  ...   
536366                           0.0                        0.0  ...   
536367                           0.0                        0.0  ...   
536368                           0.0                        0.0  ...   
536369                           0.0                        0.0  ...   

Description  wrongly coded 20713  wrongly coded 23343  wrongly coded-23343  \
InvoiceNo                                                                    
536365                       0.0                  0.0                  0.0   
536366                       0.0                  0.0                  0.0   
536367                       0.0                  0.0                  0.0   
536368                       0.0                  0.0                  0.0   
536369                       0.0                  0.0                  0.0   

Description  wrongly marked  wrongly marked 23343  \
InvoiceNo                                           
536365                  0.0                   0.0   
536366                  0.0                   0.0   
536367                  0.0                   0.0   
536368                  0.0                   0.0   
536369                  0.0                   0.0   

Description  wrongly marked carton 22804  wrongly marked. 23343 in box  \
InvoiceNo                                                                
536365                               0.0                           0.0   
536366                               0.0                           0.0   
536367                               0.0                           0.0   
536368                               0.0                           0.0   
536369                               0.0                           0.0   

Description  wrongly sold (227

For any transaction, if the quantity of an item is >=1 then its encoded as 1 (bought), else 0 (not bought)

In [10]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1

In [11]:
data = data.applymap(encode_units)
data

Description   4 PURPLE FLOCK DINNER CANDLES   50'S CHRISTMAS GIFT BAG LARGE  \
InvoiceNo                                                                     
536365                                    0                               0   
536366                                    0                               0   
536367                                    0                               0   
536368                                    0                               0   
536369                                    0                               0   
...                                     ...                             ...   
C581484                                   0                               0   
C581490                                   0                               0   
C581499                                   0                               0   
C581568                                   0                               0   
C581569                                   0                               0   

Description   DOLLY GIRL BEAKER   I LOVE LONDON MINI BACKPACK  \
InvoiceNo                                                       
536365                        0                             0   
536366                        0                             0   
536367                        0                             0   
536368                        0                             0   
536369                        0                             0   
...                         ...                           ...   
C581484                       0                             0   
C581490                       0                             0   
C581499                       0                             0   
C581568                       0                             0   
C581569                       0                             0   

Description   NINE DRAWER OFFICE TIDY   OVAL WALL MIRROR DIAMANTE   \
InvoiceNo                                                            
536365                              0                            0   
536366                              0                            0   
536367                              0                            0   
536368                              0                            0   
536369                              0                            0   
...                               ...                          ...   
C581484                             0                            0   
C581490                             0                            0   
C581499                             0                            0   
C581568                             0                            0   
C581569                             0                            0   

Description   RED SPOT GIFT BAG LARGE   SET 2 TEA TOWELS I LOVE LONDON   \
InvoiceNo                                                                 
536365                              0                                 0   
536366                              0                                 0   
536367                              0                                 0   
536368                              0                                 0   
536369                              0                                 0   
...                               ...                               ...   
C581484                             0                                 0   
C581490                             0                                 0   
C581499                             0                                 0   
C581568                             0                                 0   
C581569                             0                                 0   

Description   SPACEBOY BABY GIFT SET   TOADSTOOL BEDSIDE LIGHT   ...  \
InvoiceNo                                                        ...   
536365                             0                          0  ...   
536366                 

As we want frequently occuring items, let's ignore all the transactions where only one product is bought

In [12]:
data = data[(data > 0).sum(axis=1) >= 2]
data

Description   4 PURPLE FLOCK DINNER CANDLES   50'S CHRISTMAS GIFT BAG LARGE  \
InvoiceNo                                                                     
536365                                    0                               0   
536366                                    0                               0   
536367                                    0                               0   
536368                                    0                               0   
536372                                    0                               0   
...                                     ...                             ...   
581582                                    0                               0   
581583                                    0                               0   
581584                                    0                               0   
581585                                    0                               0   
581586                                    0                               0   

Description   DOLLY GIRL BEAKER   I LOVE LONDON MINI BACKPACK  \
InvoiceNo                                                       
536365                        0                             0   
536366                        0                             0   
536367                        0                             0   
536368                        0                             0   
536372                        0                             0   
...                         ...                           ...   
581582                        0                             0   
581583                        0                             0   
581584                        0                             0   
581585                        0                             0   
581586                        0                             0   

Description   NINE DRAWER OFFICE TIDY   OVAL WALL MIRROR DIAMANTE   \
InvoiceNo                                                            
536365                              0                            0   
536366                              0                            0   
536367                              0                            0   
536368                              0                            0   
536372                              0                            0   
...                               ...                          ...   
581582                              0                            0   
581583                              0                            0   
581584                              0                            0   
581585                              0                            0   
581586                              0                            0   

Description   RED SPOT GIFT BAG LARGE   SET 2 TEA TOWELS I LOVE LONDON   \
InvoiceNo                                                                 
536365                              0                                 0   
536366                              0                                 0   
536367                              0                                 0   
536368                              0                                 0   
536372                              0                                 0   
...                               ...                               ...   
581582                              0                                 0   
581583                              0                                 0   
581584                              0                                 0   
581585                              0                                 0   
581586                              0                                 0   

Description   SPACEBOY BABY GIFT SET   TOADSTOOL BEDSIDE LIGHT   ...  \
InvoiceNo                                                        ...   
536365                             0                          0  ...   
536366                 

### Apriori Algorithm

In [13]:
from mlxtend.frequent_patterns import apriori

In [14]:
frequent_itemset = apriori(data, min_support=0.03, use_colnames=True)

c:\Users\DELL\AppData\Local\Programs\Python\Python312\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [15]:
frequent_itemset

support                                           itemsets
0    0.051696                           (6 RIBBONS RUSTIC CHARM)
1    0.035129                  (60 CAKE CASES VINTAGE CHRISTMAS)
2    0.045529                      (60 TEATIME FAIRY CAKE CASES)
3    0.033557                   (72 SWEETHEART FAIRY CAKE CASES)
4    0.052663                       (ALARM CLOCK BAKELIKE GREEN)
..        ...                                                ...
172  0.030292  (LUNCH BAG SPACEBOY DESIGN , LUNCH BAG RED RET...
173  0.031199  (LUNCH BAG SUKI DESIGN , LUNCH BAG RED RETROSPOT)
174  0.032529  (PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN ...
175  0.033013  (PINK REGENCY TEACUP AND SAUCER, ROSES REGENCY...
176  0.031925  (WOODEN PICTURE FRAME WHITE FINISH, WOODEN FRA...

[177 rows x 2 columns]

In [16]:
frequent_itemset = frequent_itemset.sort_values('support', ascending=False).reset_index(drop=True)

In [17]:
frequent_itemset

support                                           itemsets
0    0.129875               (WHITE HANGING HEART T-LIGHT HOLDER)
1    0.116331                          (JUMBO BAG RED RETROSPOT)
2    0.100671                         (REGENCY CAKESTAND 3 TIER)
3    0.095471                                    (PARTY BUNTING)
4    0.084165                          (LUNCH BAG RED RETROSPOT)
..        ...                                                ...
172  0.030473                   (CHRISTMAS CRAFT LITTLE FRIENDS)
173  0.030473                          (CREAM HEART CARD HOLDER)
174  0.030413     (LUNCH BAG CARS BLUE, LUNCH BAG  BLACK SKULL.)
175  0.030292  (LUNCH BAG SPACEBOY DESIGN , LUNCH BAG RED RET...
176  0.030171  (LUNCH BAG SUKI DESIGN , LUNCH BAG  BLACK SKULL.)

[177 rows x 2 columns]

In [18]:
frequent_itemset['length'] = frequent_itemset['itemsets'].apply(lambda x: len(x))

In [19]:
frequent_itemset

support                                           itemsets  length
0    0.129875               (WHITE HANGING HEART T-LIGHT HOLDER)       1
1    0.116331                          (JUMBO BAG RED RETROSPOT)       1
2    0.100671                         (REGENCY CAKESTAND 3 TIER)       1
3    0.095471                                    (PARTY BUNTING)       1
4    0.084165                          (LUNCH BAG RED RETROSPOT)       1
..        ...                                                ...     ...
172  0.030473                   (CHRISTMAS CRAFT LITTLE FRIENDS)       1
173  0.030473                          (CREAM HEART CARD HOLDER)       1
174  0.030413     (LUNCH BAG CARS BLUE, LUNCH BAG  BLACK SKULL.)       2
175  0.030292  (LUNCH BAG SPACEBOY DESIGN , LUNCH BAG RED RET...       2
176  0.030171  (LUNCH BAG SUKI DESIGN , LUNCH BAG  BLACK SKULL.)       2

[177 rows x 3 columns]

### Association Rules

- Lift helps eliminate rules that occur due to chance or high popularity of items.
- Lift emphasizes how much the presence of one item increases the likelihood of another.
- A high lift value indicates a strong association, making it valuable for targeting strategies (e.g., marketing).

Due to all these factors let us find the association between the items based on the lift

In [20]:
from mlxtend.frequent_patterns import association_rules

In [21]:
association = association_rules(frequent_itemset, metric='lift', min_threshold=1, num_itemsets=len(data))

In [22]:
association.head()

antecedents                         consequents  \
0           (JUMBO BAG PINK POLKADOT)           (JUMBO BAG RED RETROSPOT)   
1           (JUMBO BAG RED RETROSPOT)           (JUMBO BAG PINK POLKADOT)   
2  (ROSES REGENCY TEACUP AND SAUCER )   (GREEN REGENCY TEACUP AND SAUCER)   
3   (GREEN REGENCY TEACUP AND SAUCER)  (ROSES REGENCY TEACUP AND SAUCER )   
4           (JUMBO BAG RED RETROSPOT)            (JUMBO STORAGE BAG SUKI)   

   antecedent support  consequent support   support  confidence       lift  \
0            0.069956            0.116331  0.047464    0.678479   5.832308   
1            0.116331            0.069956  0.047464    0.408004   5.832308   
2            0.057682            0.056473  0.042385    0.734801  13.011639   
3            0.056473            0.057682  0.042385    0.750535  13.011639   
4            0.116331            0.068323  0.042203    0.362786   5.309837   

   representativity  leverage  conviction  zhangs_metric   jaccard  certainty  \
0               1.0  0.039326    2.748400       0.890862  0.341899   0.636152   
1               1.0  0.039326    1.571032       0.937615  0.341899   0.363476   
2               1.0  0.039127    3.557807       0.979654  0.590564   0.718928   
3               1.0  0.039127    3.777361       0.978398  0.590564   0.735265   
4               1.0  0.034255    1.462109       0.918523  0.296265   0.316057   

   kulczynski  
0    0.543241  
1    0.543241  
2    0.742668  
3    0.742668  
4    0.490242

In [23]:
association.sort_values('lift', ascending=False, inplace=True)

In [24]:
association.reset_index(drop=True, inplace=True)

In [25]:
association.head(5)

antecedents                           consequents  \
0     (PINK REGENCY TEACUP AND SAUCER)     (GREEN REGENCY TEACUP AND SAUCER)   
1    (GREEN REGENCY TEACUP AND SAUCER)      (PINK REGENCY TEACUP AND SAUCER)   
2     (PINK REGENCY TEACUP AND SAUCER)    (ROSES REGENCY TEACUP AND SAUCER )   
3   (ROSES REGENCY TEACUP AND SAUCER )      (PINK REGENCY TEACUP AND SAUCER)   
4  (GARDENERS KNEELING PAD KEEP CALM )  (GARDENERS KNEELING PAD CUP OF TEA )   

   antecedent support  consequent support   support  confidence       lift  \
0            0.042264            0.056473  0.034887    0.825465  14.617093   
1            0.056473            0.042264  0.034887    0.617773  14.617093   
2            0.042264            0.057682  0.033013    0.781116  13.541798   
3            0.057682            0.042264  0.033013    0.572327  13.541798   
4            0.054235            0.045287  0.032711    0.603122  13.317793   

   representativity  leverage  conviction  zhangs_metric   jaccard  certainty  \
0               1.0  0.032500    5.405948       0.972697  0.546402   0.815019   
1               1.0  0.032500    2.505674       0.987345  0.546402   0.600906   
2               1.0  0.030575    4.305101       0.967025  0.493225   0.767717   
3               1.0  0.030575    2.239413       0.982847  0.493225   0.553454   
4               1.0  0.030254    2.405555       0.977952  0.489593   0.584296   

   kulczynski  
0    0.721619  
1    0.721619  
2    0.676721  
3    0.676721  
4    0.662709

 - We can see GREEN REGENCY TEACUP AND SAUCER and PINK REGENCY TEACUP AND SAUCER are the items that has the highest association since these two items has the highest lift value, i.e 14.6
 
- This tells us that 'PINK REGENCY TEACUP AND SAUCER' is 14.6 times more likely to be bought by the customers who buy 'GREEN REGENCY TEACUP AND SAUCER' compared to the default likelihood sale of 'PINK REGENCY TEACUP AND SAUCER’

- Since the antecedent support > consequent support, the rule that applies is {GREEN REGENCY TEACUP AND SAUCER -> PINK REGENCY TEACUP AND SAUCER}. This means that a customer has a higher tendency to buy PINK REGENCY TEACUP AND SAUCER AFTER they buy GREEN REGENCY TEACUP AND SAUCER. Not the other way around.

- The confidence level for the rule is 0.6177, which shows that out of all the transactions that contain both “GREEN REGENCY TEACUP AND SAUCER”, ~62% contain PINK REGENCY TEACUP AND SAUCER too

Similarly, we can draw other business conclusions from these association rules, which will lead in increase in revenue!